pkg/meta/sql.go

```diff
@@ -1336,45 +1340,26 @@ func (m *dbMeta) Readdir(ctx Context, inode Ino, plus uint8, entries *[]*Entry)
 		Attr:  &pattr,
 	})
 
-	rows, err := m.engine.Rows(&edge{Parent: inode})
-	if err != nil {
-		return errno(err)
-	}
-	names := make(map[Ino][]byte)
-	var inodes []string
-	for rows.Next() {
-		var e edge
-		err = rows.Scan(&e)
-		if err != nil {
-			_ = rows.Close()
-			return errno(err)
-		}
-		names[e.Inode] = []byte(e.Name)
-		inodes = append(inodes, strconv.FormatUint(uint64(e.Inode), 10))
-	}
-	_ = rows.Close()
-	if len(inodes) == 0 {
-		return 0
+	dbSession := m.engine.Table(&edge{})
+	if plus != 0 {
+		dbSession = dbSession.Table(&node{}).Join("INNER", "jfs_edge", "jfs_edge.inode=jfs_node.inode")
 	}
-
-	var n node
-	nodes, err := m.engine.Where(fmt.Sprintf("inode IN (%s)", strings.Join(inodes, ","))).Rows(&n)
-	if err != nil {
+	var nodesWithParent []nodeParent
+	if err := dbSession.Find(&nodesWithParent, &edge{Parent: inode}); err != nil {
 		return errno(err)
 	}
-	defer nodes.Close()
-	for nodes.Next() {
-		err = nodes.Scan(&n)
-		if err != nil {
-			return errno(err)
+	for _, n := range nodesWithParent {
+		entry := &Entry{
+			Inode: n.node.Inode,
+			Name:  []byte(n.edge.Name),
+			Attr:  &Attr{},
 		}
-		attr := new(Attr)
-		m.parseAttr(&n, attr)
-		*entries = append(*entries, &Entry{
-			Inode: Ino(n.Inode),
-			Name:  names[n.Inode],
-			Attr:  attr,
-		})
+		if plus != 0 {
+			m.parseAttr(&n.node, entry.Attr)
+		} else {
+			entry.Attr.Typ = n.node.Type // xorm fetches .Type is into node, actually it's from the edge table
```

Comment:
Why not use `n.edge.Type` here?

Reply:
Because xorm fills type value into `node` not `edge`😂. I left a comment here as a reminder.

Reply:
It's because we have node before edge in nodeParent. Reverse that, it just works as expected.

Reply:
Are you sure? I put node before edge because when joining two tables, their order matters according to xorm's doc. I'll try some new ways tomorrow.

Reply:
Yes, I tried, it works perfectly.

Reply:
I reversed `node` and `edge` fields, **ALSO** reversed are their orders in the `inner join` sql, so it won't break conventions in xorm.

Reply:
OK, I found it too error-prone to embed both `node` and `edge` in one struct. People may confuse where to get attributes like `inode`/`type`.

So I re-defined `namedNode` by:
```go
type namedNode struct {
	node `xorm:"extends"`
	Name string
}
```

Now `namedNode` contains everything about a node - attrs and name, without any ambiguousness. What do you think @davies